src/FsLex.Core/fslexdriver.fs

```diff
@@ -147,56 +176,82 @@ let writeUnicodeTranslationArray dfaNodes domain (writer: Writer) =
             emit Eof
-            writer.writeLine "|];"
+            writer.WriteLine "|];"
         done
 
-    writer.writeLine "    |] "
+    writer.WriteLine "    |] "
 
 let writeUnicodeActionsArray dfaNodes (writer: Writer) =
-    writer.write "let actions : uint16[] = [|"
+    writer.Write "let actions : uint16[] = [|"
     for state in dfaNodes do
         if state.Accepted.Length > 0 then
           writer.WriteUint16 (snd state.Accepted.Head)
         else
           writer.WriteUint16 sentinel
     done
-    writer.writeLine  "|]"
+    writer.WriteLine  "|]"
 
 let writeUnicodeTables lexerLibraryName domain dfaNodes (writer: Writer) =
     writeUnicodeTranslationArray dfaNodes domain writer
     writeUnicodeActionsArray dfaNodes writer
-    writer.writeLine  "let _fslex_tables = %s.%sTables.Create(trans,actions)" lexerLibraryName (match domain with | Unicode -> "Unicode" | ASCII -> "Ascii")
+    writer.WriteLine  "let _fslex_tables = %s.%sTables.Create(trans,actions)" lexerLibraryName (match domain with | Unicode -> "Unicode" | ASCII -> "Ascii")
 
 let writeRules (rules: Rule list) (perRuleData: PerRuleData) outputFileName (writer: Writer) =
-    writer.writeLine  "let rec _fslex_dummy () = _fslex_dummy() "
+    writer.WriteLine  "let rec _fslex_dummy () = _fslex_dummy() "
 
     // These actions push the additional start state and come first, because they are then typically inlined into later
     // rules. This means more tailcalls are taken as direct branches, increasing efficiency and
     // improving stack usage on platforms that do not take tailcalls.
     for (startNode, actions),(ident,args,_) in List.zip perRuleData rules do
-        writer.writeLine "// Rule %s" ident
-        writer.writeLine "and %s %s lexbuf =" ident (String.Join(" ", Array.ofList args))
-        writer.writeLine "  match _fslex_tables.Interpret(%d,lexbuf) with" startNode.Id
+        writer.WriteLine "// Rule %s" ident
+        writer.WriteLineInterface "/// Rule %s" ident
+        let arguments =
+            args
+            |> List.map (function
+                | RuleArgument.Ident ident -> ident
+                | RuleArgument.Typed(ident, typ) -> sprintf "(%s: %s)" ident typ)
+            |> String.concat " "
+
+        writer.WriteLine "and %s %s lexbuf =" ident arguments
+        
+        let signature =
+            if List.isEmpty args then
+                sprintf "val %s: lexbuf: LexBuffer<char> -> token" ident
+            else
+                args
+                |> List.map (function
+                    | RuleArgument.Ident ident ->
+                        // This is not going to lead to a valid signature file, the only workaround is that the caller will specify the type.
+                        sprintf "%s: obj" ident
+                    | RuleArgument.Typed(ident, typ) -> sprintf"%s: %s" ident typ)
+                |> String.concat " -> "
+                |> sprintf "val %s: %s -> lexbuf: LexBuffer<char> -> token" ident
```

Comment:
Minor thing: this use of currying is hard to read, would be better to be doing string interpolation here I think.

Reply:
Getting `error FS3353: Feature 'string interpolation' requires the F# library for language version 5.0 or greater.` when I tried it.
I'm guessing this is because of the low `FSharp.Core (>= 4.6.2)` dependency.
I can update this in a future PR if we can bump that version.